### PR TITLE
community: make MongodbLoader more customizable

### PR DESCRIPTION
**Pull request by pengwk**
Description:
This PR allows MongodbLoader to customize the construction of the MongoDB cursor, page_content, and metadata, as well as to decide whether to check the document count. This greatly enhances the code's extensibility.

Issue:
No

Dependencies:
No new dependencies are required for this change.

Twitter handle:
https://x.com/pengwk2

### CODE REVIEW

**Comment by pengwk** [comment]
@ccurme Do you have time to take a look at this pull request? 
Thanks!

**Review by Undertone0809 — APPROVED** [review]

File: libs/community/langchain_community/document_loaders/mongodb.py [hunk 1]
```diff
@@ -20,6 +23,10 @@ def __init__(
         *,
         filter_criteria: Optional[Dict] = None,
         field_names: Optional[Sequence[str]] = None,
+        cursor_builder: Optional[Callable] = None,
+        page_content_mapper: Optional[Callable[..., str]] = None,
+        metadata_mapper: Optional[Callable[..., Dict[str, Any]]] = None,
+        enable_total_count_check: bool = True,
```
Comment (Jibola) [inline libs/community/langchain_community/document_loaders/mongodb.py:29]: could you include docstring for this? There's an neighboring PR that's [coming down the pipe](https://github.com/langchain-ai/langchain/pull/23376/files) that would also benefit from getting the docstring info.

File: libs/community/langchain_community/document_loaders/mongodb.py [hunk 2]
```diff
@@ -59,48 +73,77 @@ def load(self) -> List[Document]:
...
+            "database": db_name,
+            "collection": collection_name,
+        }
+
+    @staticmethod
+    def page_content_default_mapper(
+        doc: Dict, field_names: Optional[Sequence[str]] = None
+    ) -> str:
+        """
+        A reasonable default function to convert a record into a "page content" string.
```
Comment (Jibola) [inline libs/community/langchain_community/document_loaders/mongodb.py:103]:
```suggestion
        Converts a record into a "page content" string.
```

File: libs/community/langchain_community/document_loaders/mongodb.py [hunk 3]
```diff
@@ -59,48 +73,77 @@ def load(self) -> List[Document]:
...
+        return collection.find(filter_criteria, projection)
+
+    @staticmethod
+    def metadata_default_mapper(
+        db_name: str,
+        collection_name: str,
+        doc: Dict,
+    ) -> Dict[str, Any]:
+        """
+        A reasonable default function to convert a doc into a "metadata" dictionary.
```
Comment (Jibola) [inline libs/community/langchain_community/document_loaders/mongodb.py:91]:
```suggestion
        Converts a doc into a "metadata" dictionary.
```

File: libs/community/langchain_community/document_loaders/mongodb.py [hunk 4]
```diff
@@ -59,48 +73,77 @@ def load(self) -> List[Document]:
...
+        else:
+            text = str(doc)
+        return text
+
     async def aload(self) -> List[Document]:
         """Load data into Document objects."""
         result = []
-        total_docs = await self.collection.count_documents(self.filter_criteria)
+        if self.enable_total_count_check:
+            total_docs = await self.collection.count_documents(self.filter_criteria)
```
Comment (Jibola) [inline libs/community/langchain_community/document_loaders/mongodb.py:130]: 🚀

**Review by Jibola — APPROVED** [review]
LGTM!

Look out for: https://github.com/langchain-ai/langchain/pull/23376/files
Depending on whose change comes in first, there will be some merge conflicts.

**Comment by pengwk** [comment]
@Jibola  Thank you for your suggestions.

**Comment by pengwk** [comment]
@Jibola @Undertone0809  Could you please let me know when this PR might be merged?

**Review by efriis — APPROVED** [review]

**Comment by ccurme** [comment]
@pengwk thanks for your patience with review.

It looks like this last merge introduced some problems. For example:
```python
    @staticmethod
    def metadata_default_mapper(
        db_name: str,
        collection_name: str,
        doc: Dict,
    ) -> Dict[str, Any]:
        """
        Converts a doc into a "metadata" dictionary.
        """
        return self._extract_fields(doc, self.metadata_names, default="")
```

`metadata_default_mapper` is no longer a staticmethod (references `self`) and doesn't use `db_name` or `collection_name`.

Would you mind resolving these issues? You can see them flagged by the linter in the view [here](https://github.com/langchain-ai/langchain/pull/25908/files).

**Comment by ccurme** [comment]
@pengwk closing for now but happy to re-open if we want to keep iterating.

**Pull request closed without merging**